src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,135 @@
+import { exec } from 'child_process';
+import { ModeName, Mode } from '../../mode/mode';
+import { existsSync, exists } from 'fs';
+import { configuration } from '../../configuration/configuration';
+import * as vscode from 'vscode';
+
+// InputMethodSwitcher change input method automatically when mode changed
+export class InputMethodSwitcher {
+  public savedIMKey = '';
+
+  public async switchInputMethod(oldMode: ModeName, newMode: ModeName) {
+    const enableAutoSwitch = configuration.autoSwitchIM.enable;
+    if (enableAutoSwitch !== true) {
+      return;
+    }
+    // when you exit from insert-like mode, save origin input method and set it to default
+    if (
+      oldMode === ModeName.Insert ||
+      oldMode === ModeName.SurroundInputMode ||
+      oldMode === ModeName.Replace
+    ) {
+      if (
+        newMode !== ModeName.Insert &&
+        newMode !== ModeName.SurroundInputMode &&
+        newMode !== ModeName.Replace
+      ) {
+        this.switchToDefaultIM();
+        return;
+      }
+    }
+    // when you enter insert-like mode, resume origin input method
+    if (
+      newMode === ModeName.Insert ||
+      newMode === ModeName.SurroundInputMode ||
+      newMode === ModeName.Replace
+    ) {
+      if (
+        oldMode !== ModeName.Insert &&
+        oldMode !== ModeName.SurroundInputMode &&
+        oldMode !== ModeName.Replace
+      ) {
+        this.resumeIM();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  public async switchToDefaultIM() {
+    const obtainIMCmd = configuration.autoSwitchIM.obtainIMCmd;
+    const rawObtainIMCmd = this.getRawCmd(obtainIMCmd);
+    if (obtainIMCmd !== '') {
+      if (existsSync(rawObtainIMCmd)) {
+        const insertIMKey = await this.execShell(obtainIMCmd);
+        if (insertIMKey !== undefined) {
+          this.savedIMKey = insertIMKey.trim();
+        }
+      } else {
+        this.showCmdNotFoundErrorMessage(rawObtainIMCmd);
+      }
+    }
+
+    const defaultIMKey = configuration.autoSwitchIM.defaultIM;
+    if (defaultIMKey !== this.savedIMKey) {
+      this.switchToIM(defaultIMKey);
+    }
+  }
+
+  // resume origin inputmethod
+  public async resumeIM() {
+    const defaultIMKey = configuration.autoSwitchIM.defaultIM;
+    if (this.savedIMKey !== defaultIMKey) {
+      this.switchToIM(this.savedIMKey);
+    }
+  }
+
+  private async switchToIM(imKey: string) {
+    let switchIMCmd = configuration.autoSwitchIM.switchIMCmd;
+    if (!switchIMCmd.includes('{im}')) {
+      this.showWrongSwitchIMCmdErrorMessage();
```

Comment:
In these error states, can we toggle the configuration to disable this plugin? Doesn't make any sense to retry again on the next command if it'll never work. 

Same comment applies to the other error cases below